trl/trainer/ppo_trainer.py

```diff
@@ -0,0 +1,611 @@
+        
+        if bs != len(queries):
+            raise ValueError(f"Batch size ({bs}) does not match number of examples ({len(queries)})")
+        
+        self._step_safety_checker(queries, responses, scores)
+
+        timing = dict()
+        t0 = time.time()
+
+        t = time.time()
+
+        logprobs, ref_logprobs, values = self.batched_forward_pass(queries, responses)
+        timing['time/ppo/forward_pass'] = time.time()-t
+
+        t = time.time()
+        rewards, non_score_reward = self.compute_rewards(scores, logprobs, ref_logprobs)
+        timing['time/ppo/compute_rewards'] = time.time()-t
+
+        t = time.time()
+        all_stats = []
+        idxs = list(range(bs))
+        for _ in range(self.config['ppo_epochs']):
+            random.shuffle(idxs)
+            for i in range(bs):
+                idx = idxs[i]
+                train_stats = self.train_minibatch(logprobs[idx].unsqueeze(0), values[idx].unsqueeze(0),
+                                                   rewards[idx].unsqueeze(0), queries[idx].unsqueeze(0),
+                                                   responses[idx].unsqueeze(0),
+                                                   torch.cat([queries[idx],responses[idx]]).unsqueeze(0))
+                all_stats.append(train_stats)
+        timing['time/ppo/optimize_step'] = time.time()-t
+
+        t = time.time()
+        train_stats = stack_dicts(all_stats)
+
+        # reshape advantages/ratios such that they are not averaged.
+        train_stats['policy/advantages'] = torch.flatten(train_stats['policy/advantages']).unsqueeze(0)
+        train_stats['policy/advantages'] = torch.nan_to_num(train_stats['policy/advantages'], WANDB_PADDING)
+        train_stats['policy/ratio'] = torch.flatten(train_stats['policy/ratio']).unsqueeze(0)
+
+        stats = self.record_step_stats(scores=scores, logprobs=logprobs, ref_logprobs=ref_logprobs,
+                                       non_score_reward=non_score_reward, train_stats=train_stats,
+                                       kl_coef=self.kl_ctl.value)
+        # Gather/Reduce stats from all processes
+        if self.is_distributed:
+            stats = self.gather_stats(stats)
+        stats = stats_to_np(stats)
+        timing['time/ppo/calc_stats'] = time.time()-t
+
+        self.kl_ctl.update(stats['objective/kl'], self.config['batch_size'])
+
+        # Log the total ppo time
+        timing['time/ppo/total'] = time.time()-t0
+        stats.update(timing)
+        return stats
+    
+    def gather_stats(self, stats):
+        """
+        Gather stats from all processes. Useful in the context of distributed training.
+
+        Args:
+            stats (dict[str, Any]): 
+            a dictionary of stats to be gathered. The stats should contain torch tensors.
+        
+        Returns:
+            stats (dict[str, Any]): 
+                a dictionary of stats with the tensors gathered.
+        """
+        import torch.distributed as dist
```

Comment:
can't we use accelerate's [`gather`](https://huggingface.co/docs/accelerate/v0.15.0/en/package_reference/accelerator#accelerate.Accelerator.gather) for this?

I think this would also be a useful thing to test a bit: so we are sure we are not screwing up the metrics gathering.